### PR TITLE
Jr 20210831 contained chained search

### PR DESCRIPTION
**Pull request by JasonRoberts-smile**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java [hunk 1]
```diff
@@ -1133,10 +1133,24 @@ private Condition createPredicateSearchParameter(@Nullable DbColumn theSourceJoi
 					break;
 				case REFERENCE:
 					for (List<? extends IQueryParameterType> nextAnd : theAndOrParams) {
-						if (theSearchContainedMode.equals(SearchContainedModeEnum.TRUE))
-							andPredicates.add(createPredicateReferenceForContainedResource(theSourceJoinColumn, theResourceName, theParamName, nextParamDef, nextAnd, null, theRequest, theRequestPartitionId));
-						else
+						if (theSearchContainedMode.equals(SearchContainedModeEnum.TRUE)) {
+							// TODO: The _contained parameter is not intended to control search chain interpretation like this.
+							//   See SMILE-2898 for details.
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:1138]: Create a related github to track the task ideally.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 2]
```diff
@@ -483,6 +483,13 @@ private void doCreateChunkedQueries(SearchParameterMap theParams, List<Long> the
 		return Optional.of(executor);
 	}
 
+	private boolean isTraverseContainedReferenceAtRoot(SearchParameterMap theParams) {
+		return myModelConfig.isIndexOnContainedResources() && theParams.values().stream()
+			.flatMap(Collection::stream)
+			.flatMap(Collection::stream)
+			.anyMatch(t -> t instanceof ReferenceParam);
+	}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:491]: I don't see how this checks what it advertises it is checking. This reads "is any SP a reference SP". Am I missing some nuance? or maybe this method should be renamed.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/ChainedContainedR4SearchTest.java [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+
+		// execute
+		myCaptureQueriesListener.clear();
+		List<String> oids = searchAndReturnUnqualifiedVersionlessIdValues(url);
+		ourLog.info(">>> " + myCaptureQueriesListener.getSelectQueriesForCurrentThread());
+
+		// validate
+		assertEquals(1L, oids.size());
+		assertThat(oids, contains(oid1.getIdPart()));
+	}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/ChainedContainedR4SearchTest.java:94]: There is definitely already a test which covers this case, and this test does not use contained values (far as I can tell). I imagine this _already_ passed when you started this branch? If so, this test should be removed. If this test failed before your implementation, then either: 
1. It should be moved to a more appropriately named test class, as this does not use contained. 
2. The test suite name should change.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/ChainedContainedR4SearchTest.java [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+
+	@Test
+	public void testShouldResolveAFourLinkChainWhereTheLastReferenceIsContained() throws Exception {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/ChainedContainedR4SearchTest.java:360]: yeah now that im reading more tests, I would rename this test class to ChainingR4SearchTest

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ContainedTest.java [hunk 5]
```diff
@@ -110,14 +107,16 @@ public void testCreateSimpleContainedResourceIndexUserDefinedId() {
...
 		SearchParameterMap map;
 
 		map = new SearchParameterMap();
 		map.add("subject", new ReferenceParam("name", "Smith"));
 		map.setSearchContainedMode(SearchContainedModeEnum.TRUE);
-		
+		map.setLoadSynchronous(true);
+		myCaptureQueriesListener.clear();
 		assertThat(toUnqualifiedVersionlessIdValues(myObservationDao.search(map)), containsInAnyOrder(toValues(id)));
+		ourLog.info(">>> " + myCaptureQueriesListener.getSelectQueriesForCurrentThread());
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ContainedTest.java:119]: Do these changes need to stay in? are they just for parity?

**Review by tadgh — APPROVED** [review]
Approved pending minor changes. Nice work!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java (hunk 1 above)
Comment (JasonRoberts-smile) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:1138]: That ticket is already in the pipeline - I don't want to preempt Calven's process.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/ChainedContainedR4SearchTest.java (hunk 4 above)
Comment (JasonRoberts-smile) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/ChainedContainedR4SearchTest.java:360]: Good call. I wanted to have a single test class that contained test methods for all of the scenarios I needed to consider, and I should have renamed the class when I started adding cases that didn't involve contained resources

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ContainedTest.java (hunk 5 above)
Comment (JasonRoberts-smile) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ContainedTest.java:119]: I thought I'd removed all of those log statements. Ken definitely wanted the setLoadSynchronous() call, so I'll leave that in.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 2 above)
Comment (JasonRoberts-smile) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:495]: Done.

**Pull request merged**